mypyc/ops_set.py

```diff
@@ -1,22 +1,109 @@
 """Primitive set ops."""
-from mypyc.ops_primitive import func_op, custom_op, simple_emit
-from mypyc.ops import object_rprimitive, bool_rprimitive, ERR_MAGIC, ERR_FALSE
+from mypyc.ops_primitive import func_op, custom_op, method_op, binary_op, simple_emit
+from mypyc.ops import (
+    object_rprimitive, bool_rprimitive, set_rprimitive, int_rprimitive, ERR_MAGIC, ERR_FALSE,
+    ERR_NEVER, EmitterInterface
+)
+from typing import List
+
 
 new_set_op = func_op(
     name='builtins.set',
     arg_types=[],
-    result_type=object_rprimitive,
+    result_type=set_rprimitive,
     error_kind=ERR_MAGIC,
-    format_str='{dest} = set()',
     emit=simple_emit('{dest} = PySet_New(NULL);')
 )
 
-# This operation is only used during set literal generation. The first argument must be a set.
-# If sets get added as special types, this will become a method_op on set rprimitives.
-set_add_op = custom_op(
-    arg_types=[object_rprimitive, object_rprimitive],
+func_op(
+    name='builtins.set',
+    arg_types=[object_rprimitive],
+    result_type=set_rprimitive,
+    error_kind=ERR_MAGIC,
+    emit=simple_emit('{dest} = PySet_New({args[0]});')
+)
+
+
+def emit_len(emitter: EmitterInterface, args: List[str], dest: str) -> None:
+    temp = emitter.temp_name()
+    emitter.emit_declaration('long long %s;' % temp)
+    emitter.emit_line('%s = PySet_GET_SIZE(%s);' % (temp, args[0]))
+    emitter.emit_line('%s = CPyTagged_ShortFromLongLong(%s);' % (dest, temp))
+
+
+func_op(
+    name='builtins.len',
+    arg_types=[set_rprimitive],
+    result_type=int_rprimitive,
+    error_kind=ERR_NEVER,
+    emit=emit_len,
+)
+
+
+binary_op(
+    op='in',
+    arg_types=[object_rprimitive, set_rprimitive],
+    result_type=bool_rprimitive,
+    error_kind=ERR_MAGIC,
+    format_str='{dest} = {args[0]} in {args[1]} :: set',
+    emit=simple_emit('{dest} = PySet_Contains({args[1]}, {args[0]});')
+)
+
+
+def emit_remove(emitter: EmitterInterface, args: List[str], dest: str) -> None:
+    temp = emitter.temp_name()
+    emitter.emit_declaration('int %s;' % temp)
+    emitter.emit_line('%s = PySet_Discard(%s, %s);' % (temp, args[0], args[1]))
+    emitter.emit_line('if (%s < 0) {' % temp)
+    emitter.emit_line('    %s = false;' % dest)
+    emitter.emit_line('} else if (%s == 0) {' % temp)
+    emitter.emit_line('    _PyErr_SetKeyError(%s);' % args[1])
+    emitter.emit_line('    %s = false;' % dest)
+    emitter.emit_line('} else {')
+    emitter.emit_line('    %s = true;' % dest)
+    emitter.emit_line('}')
```

Comment:
It might be cleaner to create a `CPySet_Discard` function in `CPy.h` and just call that, though I don't feel super strongly.
Otherwise it would be definitely cleaner to use `emitter.emit_lines`, which can cut down on the noise here.